pkg/controller/seed-controller-manager/etcdrestore/etcd_restore_controller.go

```diff
@@ -0,0 +1,267 @@
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"time"
+)
+
+const (
+	ControllerName = "kubermatic_etcd_restore_controller"
+
+	// RebuildStatefulsetFinalizer indicates that the restore is rebuilding the etcd statefulset
+	RebuildStatefulsetFinalizer = "kubermatic.io/rebuild-statefulset"
+)
+
+// Reconciler stores necessary components that are required to restore etcd backups
+type Reconciler struct {
+	log        *zap.SugaredLogger
+	workerName string
+	ctrlruntimeclient.Client
+	recorder record.EventRecorder
+}
+
+// Add creates a new etcd restore controller that is responsible for
+// managing cluster etcd restores
+func Add(
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+	numWorkers int,
+	workerName string,
+	s3Endpoint string,
+	s3BucketName string,
+	s3AccessKeyID string,
+	s3SecretAccessKey string,
+) error {
+	log = log.Named(ControllerName)
+	client := mgr.GetClient()
+
+	reconciler := &Reconciler{
+		log:        log,
+		Client:     client,
+		workerName: workerName,
+		recorder:   mgr.GetEventRecorderFor(ControllerName),
+	}
+
+	ctrlOptions := controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	}
+	c, err := controller.New(ControllerName, mgr, ctrlOptions)
+	if err != nil {
+		return err
+	}
+
+	incompleteRestorePredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			restore := e.Object.(*kubermaticv1.EtcdRestore)
+			return restore.Status.Phase != kubermaticv1.EtcdRestorePhaseCompleted
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			restore := e.ObjectNew.(*kubermaticv1.EtcdRestore)
+			return restore.Status.Phase != kubermaticv1.EtcdRestorePhaseCompleted
+		},
+	}
+
+	return c.Watch(&source.Kind{Type: &kubermaticv1.EtcdRestore{}}, &handler.EnqueueRequestForObject{}, incompleteRestorePredicates)
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	restore := &kubermaticv1.EtcdRestore{}
+	if err := r.Get(ctx, request.NamespacedName, restore); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, types.NamespacedName{Name: restore.Spec.Cluster.Name}, cluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	log = r.log.With("cluster", cluster.Name, "restore", restore.Name)
+
+	if cluster.Labels[kubermaticv1.WorkerNameLabelKey] != r.workerName {
+		return reconcile.Result{}, nil
+	}
+
```

Comment:
Does it make sense to use a cluster condition here and update it with the latest restore status?